### PR TITLE
M5-14-1: exclusion unevaluated contexts

### LINKED ISSUE #375: `M5-14-1`: Exclude unevaluated contexts
### Affected rules
 - `M5-14-1`

### Description

Exclude results in unevaluated contexts, because in practice they will never cause side-effects, and the rationale behind the rule is to report cases which may or may not cause side effects.

### Example

```cpp

```

### PR DESCRIPTION
**Pull request by knewbury01**
## Description

fixes #375 

## Change request type

 - [ ] Release or process automation (GitHub workflows, internal scripts)
 - [ ] Internal documentation
 - [ ] External documentation
 - [x] Query files (`.ql`, `.qll`, `.qls` or unit tests)
 - [ ] External scripts (analysis report or other code shipped as part of a release)

## Rules with added or modified queries

 - [ ] No rules added
 - [ ] Queries have been added for the following rules:
     - _rule number here_
 - [x] Queries have been modified for the following rules:
     - M5-14-1

## Release change checklist

A change note ([development_handbook.md#change-notes](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#change-notes)) is required for any pull request which modifies:

 - The structure or layout of the release artifacts.
 - The evaluation performance (memory, execution time) of an existing query.
 - The results of an existing query in any circumstance.

If you are only adding new rule queries, a change note is not required.

_**Author:**_ Is a change note required? 
  - [x] Yes
  - [ ] No

🚨🚨🚨
_**Reviewer:**_ Confirm that format of *shared* queries (not the .qll file, the
.ql file that imports it) is valid by running them within VS Code. 
  - [ ] Confirmed

_**Reviewer:**_ Confirm that either a change note is not required or the change note is required and has been added.
  - [ ] Confirmed

## Query development review checklist

For PRs that add new queries or modify existing queries, the following checklist should be completed by both the author and reviewer:

### Author

 - [x] Have all the relevant rule package description files been checked in?
 - [x] Have you verified that the metadata properties of each new query is set appropriately?
 - [x] Do all the unit tests contain both "COMPLIANT" and "NON_COMPLIANT" cases?
 - [x] Are the alert messages properly formatted and consistent with the [style guide](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#query-style-guide)?
 - [x] Have you run the queries on OpenPilot and verified that the performance and results are acceptable?<br />_As a rule of thumb, predicates specific to the query should take no more than 1 minute, and for simple queries be under 10 seconds. If this is not the case, this should be highlighted and agreed in the code review process._
 - [x] Does the query have an appropriate level of in-query comments/documentation?
 - [x] Have you considered/identified possible edge cases?
 - [x] Does the query not reinvent features in the standard library?
 - [x] Can the query be simplified further (not golfed!)

### Reviewer 

 - [ ] Have all the relevant rule package description files been checked in?
 - [ ] Have you verified that the metadata properties of each new query is set appropriately?
 - [ ] Do all the unit tests contain both "COMPLIANT" and "NON_COMPLIANT" cases?
 - [ ] Are the alert messages properly formatted and consistent with the [style guide](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#query-style-guide)?
 - [ ] Have you run the queries on OpenPilot and verified that the performance and results are acceptable?<br />_As a rule of thumb, predicates specific to the query should take no more than 1 minute, and for simple queries be under 10 seconds. If this is not the case, this should be highlighted and agreed in the code review process._
 - [ ] Does the query have an appropriate level of in-query comments/documentation?
 - [ ] Have you considered/identified possible edge cases?
 - [ ] Does the query not reinvent features in the standard library?
 - [ ] Can the query be simplified further (not golfed!)

### CODE REVIEW

File: cpp/autosar/src/rules/M5-14-1/RightHandOperandOfALogicalAndOperatorsContainSideEffects.ql [hunk 1]
```diff
@@ -19,10 +19,44 @@ import codingstandards.cpp.autosar
 import codingstandards.cpp.SideEffect
 import codingstandards.cpp.sideeffect.DefaultEffects
 
+/**
+ * an operator that does not evaluate its operand
+ * `decltype` also has a non evaluated operand but cannot be used in a `BinaryLogicalOperation`
+ */
+class UnevaluatedOperand extends Expr {
```
Comment (rvermeulen) [inline cpp/autosar/src/rules/M5-14-1/RightHandOperandOfALogicalAndOperatorsContainSideEffects.ql:26]: There is an opportunity for reuse using [isUnevaluated](https://codeql.github.com/codeql-standard-libraries/cpp/semmle/code/cpp/exprs/Expr.qll/predicate.Expr$Expr$isUnevaluated.0.html). 

Perhaps we can create an extended predicate in our `Expr` module that adds the `decval` case in addition to the cases covered by `isUnevaluated`.

File: cpp/autosar/test/rules/M5-14-1/test.cpp [hunk 2]
```diff
@@ -23,4 +23,14 @@ void f3(bool b) {
...
   if (b || f2()) { // COMPLIANT, f2 has local side-effects
   }
+}
+
+int g1 = 0;
+int f4() { return g1++; }
+int f5() { return 1; }
+
+void f6() {
+  if (noexcept(f5()) &&noexcept(
```
Comment (rvermeulen) [inline cpp/autosar/test/rules/M5-14-1/test.cpp:33]: Can we cover the other cases as well to ensure we properly handle those?

**Review by rvermeulen — CHANGES_REQUESTED** [review]
Some minor change request. There is some room for reuse of an existing standard library predicate that allows use to reduce our code. 

Improvement of coverage in our unit test.

Other than that all looks good!

File: cpp/autosar/test/rules/M5-14-1/test.cpp (hunk 2 above)
Comment (knewbury01) [reply cpp/autosar/test/rules/M5-14-1/test.cpp:33]: great suggestion, this let me know that actually using decval that way is also not possible! good catch!

File: cpp/autosar/src/rules/M5-14-1/RightHandOperandOfALogicalAndOperatorsContainSideEffects.ql (hunk 1 above)
Comment (knewbury01) [reply cpp/autosar/src/rules/M5-14-1/RightHandOperandOfALogicalAndOperatorsContainSideEffects.ql:26]: great catch! cant believe I missed this, oh I actually think that case is not relevant now, cannot create a testcase that would use that in the conditional (think its not possible)

File: cpp/autosar/src/rules/M5-14-1/RightHandOperandOfALogicalAndOperatorsContainSideEffects.ql [hunk 3]
```diff
@@ -19,44 +19,12 @@ import codingstandards.cpp.autosar
...
-
 from BinaryLogicalOperation op, Expr rhs
 where
   not isExcluded(op,
     SideEffects1Package::rightHandOperandOfALogicalAndOperatorsContainSideEffectsQuery()) and
   rhs = op.getRightOperand() and
   hasSideEffect(rhs) and
-  not exists(UnevaluatedOperand un | un.getOp() = rhs)
+  not rhs.(NoExceptExpr).getExpr().isUnevaluated() and
+  not rhs.(SizeofExprOperator).getExprOperand().isUnevaluated()
```
Comment (rvermeulen) [inline cpp/autosar/src/rules/M5-14-1/RightHandOperandOfALogicalAndOperatorsContainSideEffects.ql:29]: There is no need to check whether the expressions of `NoExceptExpr` and `SizeofExprOperator` are unevaluated. That is always the case.

I think it is best to reinstate `UnevalutedOperand` and replace all but the `declval` case with `isUnevaluated` on the child of `rhs`.

We can then exclude the cases by asking if `not  rhs instanceof UnevaluatedOperand`

**Review by rvermeulen — CHANGES_REQUESTED** [review]
I think we should include `isUnevaluated` as part of the `UnevaluatedOperand` and place that in the `Expr` module for reuse.

File: cpp/autosar/src/rules/M5-14-1/RightHandOperandOfALogicalAndOperatorsContainSideEffects.ql (hunk 3 above)
Comment (knewbury01) [reply cpp/autosar/src/rules/M5-14-1/RightHandOperandOfALogicalAndOperatorsContainSideEffects.ql:29]: agree! good suggestions

File: cpp/common/src/codingstandards/cpp/Expr.qll [hunk 4]
```diff
@@ -189,3 +189,17 @@ module MisraExpr {
     CValue() { isCValue(this) }
   }
 }
+
+/**
+ * an operator that does not evaluate its operand
+ */
+class UnevaluatedExprExtension extends Expr {
+  UnevaluatedExprExtension() {
+    this.getChild(_).isUnevaluated()
```
Comment (rvermeulen) [inline cpp/common/src/codingstandards/cpp/Expr.qll:198]:
```suggestion
    this.getAChild().isUnevaluated()
```

**Review by rvermeulen — CHANGES_REQUESTED** [review]
Last bit of nitpicking. Otherwise good to ship!

**Review by rvermeulen — APPROVED** [review]
:shipit:

**Pull request merged**